### PR TITLE
test: Add e2e tests for permissionless ICS

### PR DESCRIPTION
**Pull request by bermuell**
## Description

Add basic e2e tests for permissionless ICS
Change e2e tests to allow chain configurations with same ChainID

Limitation:
Current hermes version does not support two chains with same 'ChainID'. Therefore the tests Initializes two
chains with same ChainID and launches only one of them.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Provided a link to the relevant issue or specification
- [x] Reviewed "Files changed" and left comments if necessary 
- [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

File: tests/e2e/steps_permissionless_ics.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+import (
+	"time"
+
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	e2e "github.com/cosmos/interchain-security/v6/tests/e2e/testlib"
+)
+
+// stepsPermissionlessICS tests
+// - starting multiple permissionless consumer chains with the same chain ID
+// - opt-in of a validator on two different chains with the same chain ID
```
Comment (insumity) [inline tests/e2e/steps_permissionless_ics.go:12]: Mention that it tests that this "does not work" since it is not allowed. Someone reading it might think that this is possible.

File: tests/e2e/steps_permissionless_ics.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+						},
+					},
+				},
+			},
+
+			// Initialize a permissionless chain with ChainID `consu`
+			// - create the consumer chain
+			// - opt-in a validator
+			// - launch the chain
+			// - TODO (PERMISSIONLESS): 'start the chain'. Note: due to hermes we can't yet start two chains with the same ChainID as it's not yet supported
```
Comment (insumity) [inline tests/e2e/steps_permissionless_ics.go:42]: Why do we need to start the second chain? I think from testing perspective, it's okay if only one is up and running (?)
Is this a general Hermes issue (cannot connect 2 chains with the same `chainId`)?

File: tests/e2e/steps_permissionless_ics.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+							ValidatorID("bob"):   {"consu"},
+							ValidatorID("carol"): {},
+						},
+					},
+				},
+			},
+		},
+		// test chain hijacking prevention
+		[]Step{
+			// Try to change owner of chain and change deny-/allowlist
```
Comment (insumity) [inline tests/e2e/steps_permissionless_ics.go:160]: If my understanding is correct, the hijacking prevention is indirectly checked by checking that `bob` is not denylisted. Couldn't we test this by having an `ExpectedError` in `UpdateConsumerChainAction` similarly to `OptInAction`?

File: tests/e2e/steps_permissionless_ics.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+					ConsumerChain: ChainID("cons2"), // test chain "cons2" is configured with ChainID "consu"
+					SpawnTime:     uint(time.Minute * 3),
+					InitialHeight: clienttypes.Height{RevisionNumber: 0, RevisionHeight: 1},
+					TopN:          0,
+				},
+				State: State{},
+			},
+			{
+				Action: OptInAction{
+					Chain:     ChainID("cons2"),
```
Comment (insumity) [inline tests/e2e/steps_permissionless_ics.go:57]: Having this named `ChainID` while the underlying chain is is `consu` could be hard to parse. I wonder if we could rename this to something else, and introduce `ConsumerId ConsumerID` in all the actions instead of using `ChainID`.

**Review by insumity — APPROVED** [review]

File: tests/e2e/steps_permissionless_ics.go (hunk 2 above)
Comment (bermuell) [reply tests/e2e/steps_permissionless_ics.go:42]: yes, confirmed with hermes team that the identifier used in the hermes config is also used to double check that this matches the chain-id.
From an e2e testing perspective it's better IMO to have the 2nd chain also started to verify that there are no side effects, especially with issues reported by hermes. As this can't be tested due to hermes adoption to permissionless ICS the current setup is the best we can get at the moment.

File: tests/e2e/steps_permissionless_ics.go (hunk 4 above)
Comment (bermuell) [reply tests/e2e/steps_permissionless_ics.go:57]: ConsumerID would also be misleading as it is not possible at that stage to have a match with the consumer ID we get when creating the consumer chain. But I do agree to come up with something which helps to differntiate this.
Added this to #2198

File: tests/e2e/steps_permissionless_ics.go (hunk 3 above)
Comment (bermuell) [reply tests/e2e/steps_permissionless_ics.go:160]: yes, your understanding is correct. the error check would be more the whitelist approach but possible as well and might just be an additional check beside the one verifying that bob is not denylisted.

File: tests/e2e/steps_permissionless_ics.go (hunk 2 above)
Comment (MSalopek) [reply tests/e2e/steps_permissionless_ics.go:42]: I'm not sure starting the second chain is relevant.

Any side effects can be confirmed in unit and integration tests.

Per IBC rules it is fine to have 2 chains with the same chain_id. Hermes may have issues connecting 2 chains with the same chain_id, but that's not relevant for ICS.

**Review by MSalopek — APPROVED** [review]
LGTM!

We can confirm that haing 2 chains with same chain-id does not break ICS through less costly tests. Now that we have that case added to e2e we should keep it.

EDIT:
I would add steps that transition a chain from opt-in to topN via a gov proposal.

File: tests/e2e/steps_start_chains.go [hunk 5]
```diff
@@ -29,6 +32,144 @@ func stepStartProviderChain() []Step {
...
+		State: State{
+			ChainID(consumerName): ChainState{
+				ValBalances: &valBalance,
+			},
+		},
+	}
+	s = append(s, step)
+
+	// Establish IBC connection
+	steps := []Step{
```
Comment (MSalopek) [inline tests/e2e/steps_start_chains.go:147]: I would add steps that transition a chain from opt-in to topN via a gov proposal.

Comment (bermuell) [reply tests/e2e/steps_start_chains.go:147]: we have that tested already in stepsModifyChain() in `steps_partial_set_security()`
see https://github.com/cosmos/interchain-security/blob/50bbbfedeb82ee2f826569cf03bfb04d5e49e60b/tests/e2e/steps_partial_set_security.go#L2148
